### DISCUSSION TITLE
iPXE doesn't boot on Dell R720XD with QLogic 57800S NIC

### DISCUSSION THREAD

**Discussion by kaysond** [ipxe/ipxe#248] [Q&A] [answered]
I'm trying to boot netbootxyz on my R720XD. With the default image, it gets stuck at `iPXE initialising devices`.

If I use the `-snponly` version, I'm able to get past that, but then am faced with a bunch of errors:

https://ipxe.org/err/040ee1
https://ipxe.org/err/2a0860

If I hit `m` and get into the debug shell to try to set things up manually, it seems like there's an issue configuring the vlan, which gives me this error: https://ipxe.org/err/7f11e094

Does anyone know what I should try next?

**Comment by mcb30** [comment]
To simplify the test setup, could you try testing with plain upstream iPXE (rather than the version provided by netboot.xyz, which has a customised configuration and a fairly complex embedded script)?

You can build the binaries you need using `make bin-x86_64-efi/ipxe.efi bin-x86_64-efi/snponly.efi`

**Reply by kaysond** [reply]
So I built them with VLAN commands, since my network has a VLAN.

`ipxe.efi` hangs on `iPXE initialising devices...`

`snponly.efi` will start and allows me to get into the iPXE shell, but won't autoconfig, presumably because there's no VLAN tag. When I get into the shell, it won't let me create a vlan:
![image](https://user-images.githubusercontent.com/1147328/106937017-1ec00c80-66d2-11eb-9318-64a9b0b8e77b.png)

**Reply by mcb30** [reply]
Thanks for testing.  The error is coming from `efi_snp.c`.  Could you try building with `DEBUG=efi_snp` so that we can see precisely where it occurs?

**Reply by kaysond** [reply]
I figured you might ask that. I already built with `make -j 4 bin-x86_64-efi/ipxe.efi bin-x86_64-efi/snponly.efi DEBUG=efi_snp,vlan LOG_LEVEL=LOG_ALL`

Unfortunately, it didn't help much:
![image](https://user-images.githubusercontent.com/1147328/106946388-e32b3f80-66dd-11eb-82d7-6f9b4dacdb15.png)

I did discover, however, that I can create vlans on other tags!
![image](https://user-images.githubusercontent.com/1147328/106946471-fdfdb400-66dd-11eb-912f-94b907d7fb42.png)

The NIC is already configured to use vlan 17, so maybe that's why iPXE can't create it? Because net0 is already on vlan 17.

Not really sure why dhcp isn't working, from within iPXE, since obviously it worked to boot iPXE itself. I also noticed that I get some RX and TX errors:
![image](https://user-images.githubusercontent.com/1147328/106946811-706e9400-66de-11eb-8e45-a17f8bebd10c.png)

**Reply by mcb30** [reply]
> I did discover, however, that I can create vlans on other tags!
> ![image](https://user-images.githubusercontent.com/1147328/106946471-fdfdb400-66dd-11eb-912f-94b907d7fb42.png)
> 
> The NIC is already configured to use vlan 17, so maybe that's why iPXE can't create it? Because net0 is already on vlan 17.

This is plausible.  It sounds as though you potentially have something else (presumably the UEFI firmware) configured to do something on VLAN 17, and so there is already an EFI device handle present for this VLAN.  This would then prevent iPXE from installing a new handle with the same device path, which in turn prevents you from being able to use iPXE to send or receive anything on VLAN 17.

Could you build with `DEBUG=efi_driver:3` so that we can see all the devices present when iPXE first starts?

**Reply by kaysond** [reply]
Here are a couple of screenshots. I do see the vlan in there. If you need more output I can post a screen record of the whole boot.
![image](https://user-images.githubusercontent.com/1147328/106954479-b4ff2d00-66e8-11eb-9a37-8b0b027d80f0.png)
![image](https://user-images.githubusercontent.com/1147328/106954484-b92b4a80-66e8-11eb-9ccf-026673e9ec4c.png)

**Reply by kaysond** [reply]
Here's the whole boot: https://imgur.com/a/t20QZKA

**Reply by kaysond** [reply]
@mcb30 just following up to see what else I can do to debug. If it's easy and you can get me started I'm also happy to submit a PR to fix the issue.

**Reply by mcb30** [reply]
@kaysond So, in the "before disconnecting" output, I see:
```
HANDLE PcieRoot(0x0)/Pci(0x1,0x0)/Pci(0x0,0x2)/MAC(90B11C433EBD,0x0) VlanConfig opened 1x (C) by MNP Network Service Driver for PcieRoot(0x0)/Pci(0x1,0x0)/Pci(0x0,0x2)/MAC(90B11C433EBD,0x0)/Vlan(17)
```
which indicates that the `PcieRoot(0x0)/Pci(0x1,0x0)/Pci(0x0,0x2)/MAC(90B11C433EBD,0x0)/Vlan(17)` device path does already exist when iPXE starts up.

iPXE calls `DisconnectController()` to disconnect existing drivers from the NIC handle `PcieRoot(0x0)/Pci(0x1,0x0)/Pci(0x0,0x2)/MAC(90B11C433EBD,0x0)`, so that it can take over control of the device.  In the "after disconnecting" output, there are no `VlanConfig` lines, indicating that the `MNP Network Service Driver` has in theory disconnected gracefully.

However, the fact that you later get the error https://ipxe.org/7f11e094 when trying `vcreate --tag 17 net0` suggests that the device path is still present in the system, i.e. that either the `MNP Network Service Driver` or the VLAN driver (possibly called `VLAN Configuration Driver`) has not cleanly uninstalled its protocol instances.

You could try chainloading into the UEFI shell instead of iPXE, and using the `drivers`, `devices`, `devtree`, `disconnect`, `dh` and `openinfo` commands to see what happens when you attempt to `disconnect` existing drivers from the NIC handle `PcieRoot(0x0)/Pci(0x1,0x0)/Pci(0x0,0x2)/MAC(90B11C433EBD,0x0)`.

If you are able to make the `PcieRoot(0x0)/Pci(0x1,0x0)/Pci(0x0,0x2)/MAC(90B11C433EBD,0x0)/Vlan(17)` handle disappear cleanly by some sequence of UEFI shell commands, then let me know which sequence of commands was needed.

If (more likely) you find that the `PcieRoot(0x0)/Pci(0x1,0x0)/Pci(0x0,0x2)/MAC(90B11C433EBD,0x0)/Vlan(17)` device handle continues to exist even after `disconnect`, then this may suggest that one of the drivers is a candidate for our EFI veto list: https://github.com/ipxe/ipxe/blob/900f1f98d3dbb64803e427318ad61c9d7aa6f9bb/src/interface/efi/efi_veto.c#L436-L446.

**Reply by kaysond** [reply]
Sorry for being a total noob. I spent some time trying to figure out exactly how to get into the EFI shell, but the R720XD doesn't actually have a boot option for it. I guess I need a binary for it, but I couldn't find one. Could you point me in the right direction?

Interestingly, I tried booting on an R720 with a different NIC (also broadcom, though) which was also set up on that VLAN, and it worked just fine!

**Reply by NiKiZe** [reply]
You can find one at boot.ipxe.org

**Reply by kaysond** [reply]
Thanks! I ended up finding a link to one from Dell's support pages.

@mcb30 Turns out I am able to get rid of the handle by running `disconnect` on that device handle's parents. This is what its `devtree` looks like:
![image](https://user-images.githubusercontent.com/1147328/108619673-30e0c100-73db-11eb-9312-1b9b8acc6e70.png)

Running `disconnect 14D` or `disconnect 14A` didn't have any effect. Running `disconnect 149` did remove the whole tree, though. Running `disconnect CA` also does (and removes 149 with it, obviously).

I recorded my whole session, in case there's anything else that would be useful: https://drive.google.com/file/d/1YmMNY8o6QJoXFbu7yU9-U5MPJHxm4Lfd/view?usp=sharing

Please let me know if there's any other info that you need.

**Reply by mcb30** [reply]
@kaysond Thank you.  Interestingly, the base NIC device path as reported by iPXE (`PcieRoot(0x0)/Pci(0x1,0x0)/Pci(0x0,0x2)/MAC(90B11C433EBD,0x0)`, note the absence of `Vlan(17)`) does not show up **anywhere** in the output of your shell session.  It's possible that it may be the handle displayed by `devtree` as `QLogic 577xx/578xx 1 Gigabit Ethernet (BCM57800)`: could you try `dh -v` on that handle to confirm that it has the device path `PcieRoot(0x0)/Pci(0x1,0x0)/Pci(0x0,0x2)/MAC(90B11C433EBD,0x0)`?

Assuming that it does, could you please retry the disconnect tests on that handle instead?

Thanks.

**Reply by kaysond** [reply]
Yup looks like that's the one. Disconnecting it succeeds and removes all child handles, but I'm still seeing some entries in `openinfo`. Not sure if that matters
![image](https://user-images.githubusercontent.com/1147328/108758130-94fca580-74ff-11eb-86ef-611286c20dbc.png)
![image](https://user-images.githubusercontent.com/1147328/108758199-ac3b9300-74ff-11eb-89ed-50601c2c1298.png)
![image](https://user-images.githubusercontent.com/1147328/108758330-d2f9c980-74ff-11eb-98c5-295b7d7ddf16.png)
![image](https://user-images.githubusercontent.com/1147328/108758429-ee64d480-74ff-11eb-8c92-55ef962724ba.png)

**Reply by mcb30** [reply]
Thanks.  After `disconnect 149`:
- Try `dh -v 14a` to see if handle `14a` still exists (even if no longer recognised as a child of `149`)
- If not, are there any handles left in the system that have a device path of `PcieRoot(0x0)/Pci(0x1,0x0)/Pci(0x0,0x2)/MAC(90B11C433EBD,0x0)/Vlan(17)` as reported by `dh -v` (with no handle specified)?

**Reply by kaysond** [reply]
`14a` no longer exists after the `disconnect 149`. I searched `dh -v` manually and didn't see anything. Its possible I might've missed something, though. Is it possible to search it from the efi shell?

![image](https://user-images.githubusercontent.com/1147328/108761598-e6a72f00-7503-11eb-8318-f54a54653297.png)

**Reply by mcb30** [reply]
Thanks.  I think we need to start modifying code in order to be able to debug any further.

In https://github.com/ipxe/ipxe/blob/master/src/interface/efi/efi_snp.c#L1724-L1729 could you try commenting out the protocols that iPXE attempts to install and for each one check to see if `vcreate --tag 17 net0` suddenly becomes allowed?

Please note that this experimental change will create a deliberately broken build of iPXE.  You should not expect to be able to actually pass any network traffic from within iPXE, and you should definitely reboot between tests instead of expecting iPXE to exit cleanly.  The only aspect to test is whether or not `vcreate --tag 17 net0` succeeds or fails.

I would suggest starting by commenting out all but one of the protocols, and then adding them back in until `vcreate` starts failing.

**Comment by kaysond** [comment, marked as answer]
This seems to be fixed now!
